handlers/product-switch-api/src/frequencyChange.ts

```diff
@@ -0,0 +1,508 @@
+import { prettyPrint } from '@modules/prettyPrint';
+import { getProductCatalogFromApi } from '@modules/product-catalog/api';
+import type { ProductCatalog } from '@modules/product-catalog/productCatalog';
+import { ProductCatalogHelper } from '@modules/product-catalog/productCatalog';
+import { logger } from '@modules/routing/logger';
+import type { Stage } from '@modules/stage';
+import { singleTriggerDate } from '@modules/zuora/orders/orderActions';
+import type { OrderAction } from '@modules/zuora/orders/orderActions';
+import {
+	executeOrderRequest,
+	previewOrderRequest,
+} from '@modules/zuora/orders/orderRequests';
+import type {
+	CreateOrderRequest,
+	PreviewOrderRequest,
+} from '@modules/zuora/orders/orderRequests';
+import { getSubscription } from '@modules/zuora/subscription';
+import type {
+	RatePlan,
+	RatePlanCharge,
+} from '@modules/zuora/types/objects/subscription';
+import { zuoraDateFormat } from '@modules/zuora/utils/common';
+import { ZuoraClient } from '@modules/zuora/zuoraClient';
+import dayjs from 'dayjs';
+import { getCatalogBillingPeriod } from './catalogInformation';
+import {
+	frequencyChangePreviewResponseSchema,
+	frequencyChangeSwitchResponseSchema,
+	zuoraPreviewResponseSchema,
+	zuoraSwitchResponseSchema,
+} from './schemas';
+import type { ZuoraPreviewResponse, ZuoraSwitchResponse } from './schemas';
+import type {
+	FrequencyChangePreviewResponse,
+	FrequencyChangeRequestBody,
+	FrequencyChangeSwitchResponse,
+} from './schemas';
+
+/**
+ * Select the single candidate rate plan and charge eligible for a frequency change.
+ * Logic matches the handler implementation so tests can exercise preview/execute flows without duplicating filtering rules.
+ */
+export function selectCandidateSubscriptionCharge(
+	subscription: Awaited<ReturnType<typeof getSubscription>>,
+	today: Date,
+): { ratePlan: RatePlan; charge: RatePlanCharge } {
+	const candidateCharges = subscription.ratePlans
+		.filter((rp) => rp.lastChangeType !== 'Remove')
+		.flatMap((rp) =>
+			rp.ratePlanCharges.map((c) => ({ ratePlan: rp, charge: c })),
+		)
+		.filter(({ charge }) => charge.name === 'Subscription')
+		.filter(({ charge }) => charge.type === 'Recurring')
+		.filter(
+			({ charge }) =>
+				charge.effectiveStartDate <= today && charge.effectiveEndDate >= today,
```

Comment:
this will prevent them switching when they have a pending amount change (as the current dated one will be "Remove" and the "Add" one will be future dated.

Also, zuora dates are at midnight 00:00, so if we make it inclusive, we would pick up two charges when we're only covered by one
```suggestion
				charge.effectiveStartDate <= today && charge.effectiveEndDate > today,
```